tests/apps/qchem/test_dynamics.py

```diff
@@ -326,3 +332,137 @@ def test_negative_state(self, e):
+        with pytest.raises(ValueError, match="Number of samples must be at least one"):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, t, U, w, 0)
+
+    def test_invalid_time(self, c):
+        """Test if function raises a ``ValueError`` when a negative time is given."""
+        with pytest.raises(ValueError, match="Time must be zero or positive"):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, -t, U, w, ns)
+
+    def test_negative_frequency(self, c):
+        """Test if function raises a ``ValueError`` when negative frequencies are given."""
+        with pytest.raises(ValueError, match="Vibrational frequencies must be larger than zero"):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, t, U, -w, ns)
+
+    def test_zero_frequency(self, c):
+        """Test if function raises a ``ValueError`` when zero frequencies are given."""
+        with pytest.raises(ValueError, match="Vibrational frequencies must be larger than zero"):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, t, U, 0.0 * w, ns)
+
+    def test_complex_unitary(self, c):
+        """Test if function raises a ``ValueError`` when a complex unitary is given."""
+        with pytest.raises(
+            ValueError, match="The normal mode to local mode transformation matrix must be real"
+        ):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, t, 1.0j * U, w, ns)
+
+    def test_invalid_loss(self, c):
+        """Test if function raises a ``ValueError`` when the loss parameter is specified outside
+        of range."""
+        with pytest.raises(
+            ValueError, match="Loss parameter must take a value between zero and one"
+        ):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha, t, U, w, ns, -1)
+
+    def test_invalid_mode(self, c):
+        """Test if function raises a ``ValueError`` when the number of displacement parameters and the
+        number of modes in the normal-to-local transformation matrix are different."""
+        with pytest.raises(
+            ValueError,
+            match="Number of displacement parameters and the number of modes in the normal-to-local",
+        ):
+            alpha, t, U, w, ns = c
+            dynamics.sample_coherent(alpha + [0], t, U, w, ns)
+
+
+    def test_no_loss(self, monkeypatch, c):
+        """Test if function correctly creates the SF program for circuits without loss."""
+
+        def save_hist(*args, **kwargs):
+            call_history.append(args[1])
+            return sf.engine.Result
+
+        call_history = []
+        with monkeypatch.context() as m:
+            m.setattr(sf.engine.Result, "samples", np.array([[0]]))
+            m.setattr(sf.LocalEngine, "run", save_hist)
+            dynamics.sample_coherent(*c)
+
+        assert not all([isinstance(op, sf.ops.LossChannel) for op in call_history[0].circuit])
```

Comment:
Not crucial, but I think these tests are identical to the `sample_fock` tests, it might be possible to loop over the two functions using `@pytest.mark.parametrize("func", [dynamics.sample_fock, dynamics.sample_coherent])` and save the repetition.

Reply:
This would also save on testing for the `sample_tmsv` function, but of course is a bit of work now.